### PR TITLE
warm start search, add showLive=0

### PR DESCRIPTION
**Pull request by ccstan99**
1. After loading, warm up model with a dummy search to speed up later searches
2. Experimenting with debounce values, was changed from 400 to 500 on 1/22
3. Add showLive=0 for semantic search endpoint so suggesting new question only show non-canonically answered questions where status is NOT "Live on site"

### CODE REVIEW

File: app/components/search.tsx [hunk 1]
```diff
@@ -79,7 +79,8 @@ export default function Search({onSiteAnswersRef, openQuestionTitles, onSelect}:
     }
   }
 
-  const handleChange = debounce(searchFn, 500)
+  // const handleChange = debounce(searchFn, 300)
+  const handleChange = searchFn
```
Comment (Aprillion) [inline app/components/search.tsx:83]: have you tested in Firefox and on mobile browsers? debounce was added because search was still running for out-of-date queries while typing additional letters and then it took up to 2 seconds of blinking results until the final results were shown

to search without delay, you can always press Enter

not sure why I changed it to 500ms at the same time then adding the Enter trigger,, probably left over by mistake while testing - feel free to use 300 or even 100ms, but I am not sure about the zero debounce

**Comment by tayler6000** [comment]
@ccstan99 @Aprillion Would it be fair to say this PR should be linked to #175 ?

File: app/components/search.tsx (hunk 1 above)
Comment (ccstan99) [reply app/components/search.tsx:83]: Sorry, I had this PR as a draft because it's not ready for review yet. Probably should have just dropped a message in discord. I have a demo at https://stampy-ui.ccstan99.workers.dev/ so we can test it on Firefox and mobile. I didn't see the blinking results on Firefox when I tested it on my machine but don't expect it to have magically fixed for people who were seeing that either.

Comment (ccstan99) [reply app/components/search.tsx]: OK, I set the debounce to 100ms for now.

**Comment by ccstan99** [comment]
I also added userQuery along with the TF results as a start to addressing #67

File: app/routes/questions/add.tsx [hunk 2]
```diff
@@ -5,7 +5,7 @@ import {redirect} from '@remix-run/cloudflare'
 import {addQuestion, loadAllQuestions, fetchJsonList, RelatedQuestions} from '~/server-utils/stampy'
 
 const getRelated = async (question: string): Promise<RelatedQuestions> => {
-  const url = `${NLP_SEARCH_ENDPOINT}/api/search?query=${question}`
+  const url = `${NLP_SEARCH_ENDPOINT}/api/search?query=${question}&showLive=0`
```
Comment (mruwnik) [inline app/routes/questions/add.tsx:8]: shouldn't these be for all questions, rather than just unpublished?

File: app/components/search.tsx [hunk 3]
```diff
@@ -23,12 +23,11 @@ type SearchResult = Question & {
   url?: string
 }
 
-type WorkerMessage =
```
Comment (mruwnik) [inline app/components/search.tsx:26]: [NIT] it seems a bit iffy to have this as two separate types

Comment (ccstan99) [reply app/components/search.tsx:26]: Good point! I was so fixated on getting rid of the TypeScript errors that I didn't think through the fundamentals. Well, I tried this...
```
type WorkerMessage =
  | { status: string; numQs: number }
  | { searchResults: {title: string; pageid: string; score: number}[]; userQuery: string }
```
but it's still giving me these errors:
```
app/components/search.tsx:45:16 - error TS2339: Property 'status' does not exist on type 'WorkerMessage'.
  Property 'status' does not exist on type '{ searchResults: { title: string; pageid: string; score: number; }[]; userQuery: string; }'.
45       if (data.status === 'ready') {
                  ~~~~~~
app/components/search.tsx:49:16 - error TS2339: Property 'searchResults' does not exist on type 'WorkerMessage'.
  Property 'searchResults' does not exist on type '{ status: string; numQs: number; }'.
49       if (data.searchResults) {
                  ~~~~~~~~~~~~~
app/components/search.tsx:50:31 - error TS2339: Property 'searchResults' does not exist on type 'WorkerMessage'.
  Property 'searchResults' does not exist on type '{ status: string; numQs: number; }'.
50         setSearchResults(data.searchResults)
```
So I'm obviously missing something. To be honest, my understanding of TypeScript is purely by trial and error. How would you suggest to fix this properly?

File: app/routes/questions/add.tsx (hunk 2 above)
Comment (ccstan99) [reply app/routes/questions/add.tsx]: Oops, I think I might have clobbered the work you did yesterday since I created this branch & made these changes earlier but didn't have time to continue working on this until now. I thought this function was for when the user is suggesting a new question. If that's the case, they should be requested only unpublished / non-live questions. Otherwise presumably, they should have picked a live / published answered from the previous screen. Anyway, I removed it.

Comment (mruwnik) [reply app/routes/questions/add.tsx]: won't it need `status=all`? i.e. I forgot about this url, so it's good that you found it :D

File: app/components/search.tsx (hunk 3 above)
Comment (mruwnik) [reply app/components/search.tsx:26]: It think something like this should work:
```
type WorkerMessage = {
  searchResults?: {title: string; pageid: string; score: number}[]
  status: string
  numQs?: number
}
```

File: app/routes/questions/add.tsx (hunk 2 above)
Comment (Aprillion) [reply app/routes/questions/add.tsx]: @mruwnik I also thought that this is for unpublished suggestions,, why do we want all, please?

Comment (mruwnik) [reply app/routes/questions/add.tsx]: this is what gets put into the possible duplicates column of https://coda.io/d/AI-Safety-Info_dfau7sl2hmG/Incoming-questions_su7CE#_luTjm
The unpublished questions for the dialog thingy that are to be bumped, are in `app/routes/questions/search.tsx`

Comment (mruwnik) [reply app/routes/questions/add.tsx]: There really should be some kind of central place/abstraction for all this :/

File: app/components/search.tsx (hunk 3 above)
Comment (ccstan99) [reply app/components/search.tsx:26]: Ah, that did it, thanks! Aprillion tried explaining to me what the ? meant but I guess it didn't stick.

File: app/routes/questions/add.tsx (hunk 2 above)
Comment (ccstan99) [reply app/routes/questions/add.tsx]: I think the user should only see and request unpublished questions in the dialog box. What do you think, @plex? I thought the return list was getting filtered out afterward and that's why there weren't always 5 listed.

Comment (mruwnik) [reply app/routes/questions/add.tsx]: that was the previous approach before you added filtering. Currently, there are 3 places that use search:
* the main search bar that uses tensor flow
* the dialog with unpublished questions that can be boosted -> this is in `app/routes/questions/search.tsx`
* an additional call to semantic search for all similar questions, when adding a new question -> this one here

Comment (ccstan99) [reply app/routes/questions/add.tsx]: It's way past my bedtime and I'm delirious... I'm not fully following the difference between the last 2. So, are the results from this list ever presented to the user? Or is it just one additional check before adding it to Coda? Anyway, I have an impending work deadline, so I'll have to leave this as is for now. If you need this fixed or merged sooner, feel free to finish it off. Otherwise unfortunately, I won't be able to get back to this again until later in the week. Thanks!

Comment (mruwnik) [reply app/routes/questions/add.tsx]: It's not even a check. The Incoming questions table in coda has a column for similar questions that are possible duplicates, to make it easier to decide how to handle new questions. This column is populated from semantic search, over the whole set of questions. Coda has a button that will go through the incoming questions and update the possible duplicates with values from semantic search. But the UI, when creating a new question, can also fill that field. So when a new question is created, the UI will additionally fetch a bunch of similar questions and set the possible duplicates directly.

This is a long-winded way of saying that no, the user will never see this list. It's all handled in the backend and only used when writing stuff to Coda

Comment (ccstan99) [reply app/routes/questions/add.tsx]: I knew you were doing some backend magic with Coda, so I'll trust that I should just stay out of your way. Thanks for clarifying so much for though! Otherwise, I totally missed that point. Perhaps you might consider checking if the score is at least above some threshold (i.e. >0.7?) to avoid cluttering the Coda column with rather dissimilar questions?

**Review by mruwnik — DISMISSED** [review]

File: app/routes/questions/add.tsx (hunk 2 above)
Comment (mruwnik) [reply app/routes/questions/add.tsx]: The threshold can be set as a parameter in the Coda formula. Currently it's 0.3, which is a bit low, but the idea was to have 5 for each item, and the reviewer can decide what to do with them.

File: app/components/search.tsx (hunk 3 above)
Comment (mruwnik) [reply app/components/search.tsx:26]: yup, typescript is annoying in that way :D

Comment (Aprillion) [reply app/components/search.tsx:26]: typescript is nice, when all types are shared ... but the worker is native JavaScript file (executed as-is without compilation in the browser for now), so the TS type must match what the worker actually sends and there is no TS error if we forget to synchronize both

feel free to refactor to compile the workef rom TS you want to make it less annoying 😸

Comment (Aprillion) [reply app/components/search.tsx:26]: actually, the type doesn't match - following lines post message from the worker:
```js
23: self.postMessage({status: 'ready', numQs: questions.length})
39: self.postMessage({searchResults: []})
68: self.postMessage({searchResults, userQuery})
```

the correct type for that is:
```ts
type WorkerMessage = {
  status: 'ready'
  numQs: number
} | {
  searchResults: {title: string, pageid: string, score: number}[]
  userQuery?: string
}
```

and fixing the type will probably produce TS errors in this file => @ccstan99 let me know if you want me to fix all TS thingies myself

**Review by Aprillion — CHANGES_REQUESTED** [review]

File: app/components/search.tsx (hunk 3 above)
Comment (ccstan99) [reply app/components/search.tsx:26]: Yes, pretty please! I'd love to learn TS (and a bunch of other things) properly some day. I can definitely see its potential for beauty and power. Unfortunately, I don't have the bandwidth today.

Comment (ccstan99) [reply app/components/search.tsx:26]: @Aprillion Should I wait for you to take care of the TS thingies before merging? Or should we merge first and cleanup later? Otherwise, I'll need to find some time to go back to my trial and error approach with TS. Would you mind sending me that link that explains how those ? work in TS again? I forgot what they're even called. I remember it made sense initially, but unfortunately, old age & memory don't go well together.

**Review by Aprillion — APPROVED** [review]
